dashboard/src/actions/datasetListActions.js

```diff
@@ -1,27 +1,74 @@
 import * as TYPES from "./types";
 
 import API from "../utils/axiosInstance";
+import { DANGER } from "assets/constants/toastConstants";
+import { showToast } from "./toastActions";
 import { uriTemplate } from "utils/helper";
 
-export const fetchPublicDatasets = () => async (dispatch, getState) => {
+export const fetchPublicDatasets = (page) => async (dispatch, getState) => {
   try {
     dispatch({ type: TYPES.LOADING });
     const endpoints = getState().apiEndpoint.endpoints;
+    const { offset, limit, filter, searchKey, perPage } =
+      getState().datasetlist;
+    let publicData = [...getState().datasetlist.publicData];
+    const params = new URLSearchParams();
+    params.append("metadata", "dataset.uploaded");
+    params.append("access", "public");
+    params.append("offset", offset);
+    params.append("limit", limit);
+
+    if (searchKey) {
+      params.append("name", searchKey);
+    }
+    if (filter.startDate instanceof Date && !isNaN(filter.startDate)) {
+      params.append("start", new Date(filter.startDate).toUTCString());
+    }
+    if (filter.endDate instanceof Date && !isNaN(filter.endDate)) {
+      params.append("end", new Date(filter.endDate).toUTCString());
+    }
+
     const response = await API.get(
       uriTemplate(endpoints, "datasets_list", {}),
-      { params: { metadata: "dataset.uploaded", access: "public" } }
+      { params }
     );
+
     if (response.status === 200 && response.data) {
+      const startIdx = (page - 1) * perPage;
+
+      if (publicData.length === 0) {
+        publicData = new Array(response.data.total);
+      }
+      publicData.splice(
+        startIdx,
+        response.data.results.length,
+        ...response.data.results
+      );
+
+      dispatch({
+        type: TYPES.GET_PUBLIC_DATASETS,
+        payload: [...publicData],
+      });
+      // in case of last page, next_url is empty
+      if (response.data.next_url) {
+        const urlSearchParams = new URLSearchParams(response.data.next_url);
+        const params = Object.fromEntries(urlSearchParams.entries());
+
+        dispatch({
+          type: TYPES.SET_PAGE_OFFSET,
+          payload: Number(params?.offset),
+        });
```

Comment:
> But a better question is, given that the `/datasets/list` method returns the `total`, it seems like it might be friendly for it also to return the current (or next) offset, as well...so that the client doesn't have to reverse-engineer the `next_url` in order to confirm where it is in the sequence. (Thoughts, @dbutenhof ?)

When the pagination interface was designed, we were thinking of sequential access, and the `next_url` was convenient. For random access, maybe we should capture the query parameter object (which we're really creating anyway) rather than (just) stringifying it. E.g.,

```json
{
  "next_url": "http://pbench.app.example.com/api/v1/datasets?mine&offset=30&limit=10"
  "next_params": {
    "mine": "true",
    "offset": 30,
    "limit": 10
  },
  "results": [ ... ],
  "total": 60
}
```

I actually kinda like that, and I might code it up ...

Reply:
That seems good.  My only question is, should the extra data be describing the current result or the next?

My instinct is that the code looking at the response probably knows (or could easily know) what the filter parameters were and what the limit was; however, knowing what the offset was is harder (i.e., this isn't static, and it would have to be tracked).  Now, it's true that what you're considering above can be used to calculate the offset for the current response (simply by subtracting `1`), but....

I guess the real question is, how do we expect the additional fields to be used?  Do we really expect the client to create a new URL from them?  (Isn't `"next_url"` a sufficient basis for that?)  Or, do we expect the client to use the information for some other purpose?...in which case, perhaps what we return should be better suited for the intended/expected use.

Reply:
We can't predict what the client will want next, but giving them what they just gave us doesn't seem very useful either. Right now Varshini is just parsing out `next_url`, which is fine, but you wanted that to be easier and more general; and while that makes sense, there's nothing we can reasonably give except "now" or "next" and to me just returning the individual query parameters that are summarized in `next_url` (that is, the client doesn't need to parse it out) seems most logical of the possible alternatives.

Reply:
OK, I totally missed the fact that `fetchPublicDatasets()` isn't using the `next_url` value directly...instead it's reassembling it from pieces, one of which the Dashboard parses out of `next_url`.  Based on that, the Server should be returning`"next_params"` _instead_ of `"next_url"`....  (But, yes, I think the "sequential access case" is important to provide for, even if the Dashboard doesn't use it, so we should presumably return both.)

But, if our expectation is that the Dashboard will essentially be doing random access, then having `"next_url"` return all the information required to build a query for the next page of data isn't really on-target.  Instead, it would make more sense to return the information required to re-build the current query, with the understanding that the client will extrapolate from that (and the `total`) to build the query that the user wants next.

(I'm kind of frustrated by the fact that the code seems to be maintaining two, separate notions of what the current/next position is:  `fetchPublicDatasets()` receives a `page` parameter, but it uses the `offset` value from the Redux store to build the request URL...I understand that one of those is in units of "pages" and the other is in units of "results", but I'm concerned that they won't stay in sync if they are maintained separately.)